### PR TITLE
Port structural changes from Multi-Tab

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This PR contains a port of the structural changes from the Multi-Tab PR (https://github.com/firebase/firebase-js-sdk/pull/1000)

It's might be worth discussing just getting rid of the TargetIdGenerator at this point...

### CODE REVIEW

**Comment by mikelehen** [comment]
btw, if you merge the latest master there should be a slightly higher chance CI will pass.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/ViewSnapshot.java [hunk 1]
```diff
@@ -53,6 +55,26 @@ public ViewSnapshot(
...
+      viewChanges.add(DocumentViewChange.create(DocumentViewChange.Type.ADDED, doc));
+    }
+    return new ViewSnapshot(
+        query,
+        documents,
+        DocumentSet.emptySet(query.comparator()),
+        viewChanges,
+        fromCache,
+        mutatedKeys,
+        true);
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/ViewSnapshot.java]: rumble

[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/ViewSnapshot.java]: Note that this changed the unit tests quite a bit, as Web doesn't verify the `excludeMetadataChanges` flag.

**Pull request merged**